File: mem0/llms/aws_bedrock.py
import json
import os
import re
from typing import Any, Dict, List, Optional

try:
    import boto3
except ImportError:
    raise ImportError("The 'boto3' library is required. Please install it using 'pip install boto3'.")

from mem0.configs.llms.base import BaseLlmConfig
from mem0.llms.base import LLMBase


PROVIDERS = ["ai21", "amazon", "anthropic", "cohere", "meta", "mistral", "stability", "writer"]


def extract_provider(model: str) -> str:
    for provider in PROVIDERS:
        if re.search(rf"\b{re.escape(provider)}\b", model):
            return provider
    raise ValueError(f"Unknown provider in model: {model}")


class AWSBedrockLLM(LLMBase):
    def __init__(self, config: Optional[BaseLlmConfig] = None):
        super().__init__(config)

        if not self.config.model:
            self.config.model = "anthropic.claude-3-5-sonnet-20240620-v1:0"

        # Get AWS config from environment variables or use defaults
        aws_access_key = os.environ.get("AWS_ACCESS_KEY_ID", "")
        aws_secret_key = os.environ.get("AWS_SECRET_ACCESS_KEY", "")
        aws_region = os.environ.get("AWS_REGION", "us-west-2")

        # Check if AWS config is provided in the config
        if hasattr(self.config, "aws_access_key_id"):
            aws_access_key = self.config.aws_access_key_id
        if hasattr(self.config, "aws_secret_access_key"):
            aws_secret_key = self.config.aws_secret_access_key
        if hasattr(self.config, "aws_region"):
            aws_region = self.config.aws_region

        self.client = boto3.client(
            "bedrock-runtime",
            region_name=aws_region,
            aws_access_key_id=aws_access_key if aws_access_key else None,
            aws_secret_access_key=aws_secret_key if aws_secret_key else None,
        )

        self.model_kwargs = {
            "temperature": self.config.temperature,
            "max_tokens_to_sample": self.config.max_tokens,
            "top_p": self.config.top_p,
        }

    def _format_messages(self, messages: List[Dict[str, str]]) -> str:
        """
        Formats a list of messages into the required prompt structure for the model.

        Args:
            messages (List[Dict[str, str]]): A list of dictionaries where each dictionary represents a message.
                                            Each dictionary contains 'role' and 'content' keys.

        Returns:
            str: A formatted string combining all messages, structured with roles capitalized and separated by newlines.
        """

        formatted_messages = []
        for message in messages:
            role = message["role"].capitalize()
            content = message["content"]
            formatted_messages.append(f"\n\n{role}: {content}")

        return "\n\nHuman: " + "".join(formatted_messages) + "\n\nAssistant:"

    def _parse_response(self, response, tools) -> str:
        """
        Process the response based on whether tools are used or not.

        Args:
            response: The raw response from API.
            tools: The list of tools provided in the request.

        Returns:
            str or dict: The processed response.
        """
        if tools:
            processed_response = {"tool_calls": []}

            if response["output"]["message"]["content"]:
                for item in response["output"]["message"]["content"]:
                    if "toolUse" in item:
                        processed_response["tool_calls"].append({
                            "name": item["toolUse"]["name"],
                            "arguments": item["toolUse"]["input"],
                        })

            return processed_response

        response_body = response.get("body").read().decode()
        response_json = json.loads(response_body)
        return response_json.get("content", [{"text": ""}])[0].get("text", "")

    def _prepare_input(
        self,
        provider: str,
        model: str,
        prompt: str,
        model_kwargs: Optional[Dict[str, Any]] = {},
    ) -> Dict[str, Any]:
        """
        Prepares the input dictionary for the specified provider's model by mapping and renaming
        keys in the input based on the provider's requirements.

        Args:
            provider (str): The name of the service provider (e.g., "meta", "ai21", "mistral", "cohere", "amazon").
            model (str): The name or identifier of the model being used.
            prompt (str): The text prompt to be processed by the model.
            model_kwargs (Dict[str, Any]): Additional keyword arguments specific to the model's requirements.

        Returns:
            Dict[str, Any]: The prepared input dictionary with the correct keys and values for the specified provider.
        """

        input_body = {"prompt": prompt, **model_kwargs}

        provider_mappings = {
            "meta": {"max_tokens_to_sample": "max_gen_len"},
            "ai21": {"max_tokens_to_sample": "maxTokens", "top_p": "topP"},
            "mistral": {"max_tokens_to_sample": "max_tokens"},
            "cohere": {"max_tokens_to_sample": "max_tokens", "top_p": "p"},
        }

        if provider in provider_mappings:
            for old_key, new_key in provider_mappings[provider].items():
                if old_key in input_body:
                    input_body[new_key] = input_body.pop(old_key)

        if provider == "cohere" and "cohere.command-r" in model:
            input_body["message"] = input_body.pop("prompt")

        if provider == "amazon":
            input_body = {
                "inputText": prompt,
                "textGenerationConfig": {
                    "maxTokenCount": self.model_kwargs["max_tokens_to_sample"]
                    or self.model_kwargs["max_tokens"]
                    or 5000,
                    "topP": self.model_kwargs["top_p"] or 0.9,
                    "temperature": self.model_kwargs["temperature"] or 0.1,
                },
            }
            input_body["textGenerationConfig"] = {
                k: v for k, v in input_body["textGenerationConfig"].items() if v is not None
            }

        return input_body

    def _convert_tool_format(self, original_tools):
        """
        Converts a list of tools from their original format to a new standardized format.

        Args:
            original_tools (list): A list of dictionaries representing the original tools, each containing a 'type' key and corresponding details.

        Returns:
            list: A list of dictionaries representing the tools in the new standardized format.
        """
        new_tools = []

        for tool in original_tools:
            if tool["type"] == "function":
                function = tool["function"]
                new_tool = {
                    "toolSpec": {
                        "name": function["name"],
                        "description": function["description"],
                        "inputSchema": {
                            "json": {
                                "type": "object",
                                "properties": {},
                                "required": function["parameters"].get("required", []),
                            }
                        },
                    }
                }

                for prop, details in function["parameters"].get("properties", {}).items():
                    new_tool["toolSpec"]["inputSchema"]["json"]["properties"][prop] = details

                new_tools.append(new_tool)

        return new_tools

    def generate_response(
        self,
        messages: List[Dict[str, str]],
        response_format=None,
        tools: Optional[List[Dict]] = None,
        tool_choice: str = "auto",
    ):
        """
        Generate a response based on the given messages using AWS Bedrock.

        Args:
            messages (list): List of message dicts containing 'role' and 'content'.
            tools (list, optional): List of tools that the model can call. Defaults to None.
            tool_choice (str, optional): Tool choice method. Defaults to "auto".

        Returns:
            str: The generated response.
        """

        if tools:
            # Use converse method when tools are provided
            messages = [
                {
                    "role": "user",
                    "content": [{"text": message["content"]} for message in messages],
                }
            ]
            inference_config = {
                "temperature": self.model_kwargs["temperature"],
                "maxTokens": self.model_kwargs["max_tokens_to_sample"],
                "topP": self.model_kwargs["top_p"],
            }
            tools_config = {"tools": self._convert_tool_format(tools)}

            response = self.client.converse(
                modelId=self.config.model,
                messages=messages,
                inferenceConfig=inference_config,
                toolConfig=tools_config,
            )
        else:
            # Use invoke_model method when no tools are provided
            prompt = self._format_messages(messages)
            provider = extract_provider(self.config.model)
            input_body = self._prepare_input(provider, self.config.model, prompt, model_kwargs=self.model_kwargs)
            body = json.dumps(input_body)

            if provider == "anthropic" or provider == "deepseek":
                input_body = {
                    "messages": [{"role": "user", "content": [{"type": "text", "text": prompt}]}],
                    "max_tokens": self.model_kwargs["max_tokens_to_sample"] or self.model_kwargs["max_tokens"] or 5000,
                    "temperature": self.model_kwargs["temperature"] or 0.1,
                    "top_p": self.model_kwargs["top_p"] or 0.9,
                    "anthropic_version": "bedrock-2023-05-31",
                }

                body = json.dumps(input_body)

                response = self.client.invoke_model(
                    body=body,
                    modelId=self.config.model,
                    accept="application/json",
                    contentType="application/json",
                )
            else:
                response = self.client.invoke_model(
                    body=body,
                    modelId=self.config.model,
                    accept="application/json",
                    contentType="application/json",
                )

        return self._parse_response(response, tools)

File: mem0-ts/src/oss/src/llms/aws_bedrock.ts
import { LLM, LLMResponse } from "./base";
import { LLMConfig, Message } from "../types";

/**
 * Providers recognised in Bedrock model identifiers, mirroring the Python
 * provider's `PROVIDERS` list (`mem0/llms/aws_bedrock.py`).
 */
const PROVIDERS = [
  "ai21",
  "amazon",
  "anthropic",
  "cohere",
  "meta",
  "mistral",
  "stability",
  "writer",
  "deepseek",
  "gpt-oss",
  "perplexity",
  "snowflake",
  "titan",
  "command",
  "j2",
  "llama",
  "minimax",
];

/**
 * Extract the model-family provider from a Bedrock model id
 * (e.g. `anthropic.claude-3-sonnet-...` -> `anthropic`).
 */
export function extractProvider(model: string): string {
  for (const provider of PROVIDERS) {
    const re = new RegExp(
      `\\b${provider.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\b`,
    );
    if (re.test(model)) return provider;
  }
  throw new Error(`Unknown provider in model: ${model}`);
}

/**
 * AWS Bedrock fields (awsRegion / awsAccessKeyId / awsSecretAccessKey /
 * awsSessionToken / client) now live on the shared `LLMConfig`, so the
 * provider is configurable through the standard typed `Memory` config path.
 */
type AWSBedrockConfig = LLMConfig;

/**
 * AWS Bedrock LLM provider for the TypeScript OSS SDK.
 *
 * Mirrors `mem0/llms/aws_bedrock.py`. Uses the Bedrock **Converse API**
 * (`ConverseCommand`) by default, which provides a uniform message/tool interface
 * across the Anthropic / Amazon (Nova) / Meta / Mistral / Cohere model families.
 * Opt into the legacy `InvokeModelCommand` path via `bedrockApi: "invoke_model"`
 * for older models not reachable through Converse (#6023).
 *
 * The `@aws-sdk/client-bedrock-runtime` dependency is loaded on first use via
 * dynamic `import()` so the package stays optional. Credentials resolve via the
 * standard AWS chain unless provided explicitly in config.
 */
/**
 * Bedrock API path. `converse` (default) uses the uniform Converse API.
 * `invoke_model` uses legacy InvokeModelCommand for older models (#6023).
 */
export type BedrockApi = "converse" | "invoke_model";

interface BedrockSDK {
  BedrockRuntimeClient: new (config: Record<string, any>) => any;
  ConverseCommand: new (input: Record<string, any>) => any;
  InvokeModelCommand: new (input: Record<string, any>) => any;
}

export class AWSBedrockLLM implements LLM {
  private model: string;
  private provider: string;
  private temperature: number;
  private maxTokens: number;
  private topP?: number;
  private api: BedrockApi;
  private clientConfig: Record<string, any>;
  private clientOverride?: any;
  private sdkPromise?: Promise<BedrockSDK>;
  private clientPromise?: Promise<any>;

  constructor(config: AWSBedrockConfig = {}) {
    this.model =
      (typeof config.model === "string" && config.model) ||
      "anthropic.claude-3-5-sonnet-20240620-v1:0";
    this.provider = extractProvider(this.model);
    this.temperature = config.temperature ?? 0.1;
    this.maxTokens = config.maxTokens ?? 2000;
    this.topP = config.topP;
    this.api =
      config.bedrockApi === "invoke_model" ? "invoke_model" : "converse";

    const region =
      config.awsRegion ||
      process.env.AWS_REGION ||
      process.env.AWS_DEFAULT_REGION;
    const clientConfig: Record<string, any> = {};
    if (region) clientConfig.region = region;
    if (config.awsAccessKeyId && config.awsSecretAccessKey) {
      clientConfig.credentials = {
        accessKeyId: config.awsAccessKeyId,
        secretAccessKey: config.awsSecretAccessKey,
        ...(config.awsSessionToken && {
          sessionToken: config.awsSessionToken,
        }),
      };
    }

    this.clientConfig = clientConfig;
    this.clientOverride = config.client;
  }

  /**
   * Load the optional AWS SDK on first use.
   *
   * This MUST be a dynamic `import()`, never `require()`: tsup/esbuild rewrite
   * `require()` in the published ESM bundle (`dist/oss/index.mjs`) into a
   * `__require` shim that throws `Dynamic require of "..." is not supported`,
   * so every ESM consumer would hit a dead provider even with the SDK installed.
   */
  private async getSDK(): Promise<BedrockSDK> {
    if (!this.sdkPromise) {
      // Peer dep may be absent in the typecheck environment.
      // @ts-expect-error optional peer @aws-sdk/client-bedrock-runtime
      this.sdkPromise = import("@aws-sdk/client-bedrock-runtime").then(
        (sdk) => sdk as unknown as BedrockSDK,
        (err) => {
          // Let a later call retry rather than caching the rejection forever.
          this.sdkPromise = undefined;
          const detail = err instanceof Error ? err.message : String(err);
          throw new Error(
            "The '@aws-sdk/client-bedrock-runtime' package is required to use the AWS Bedrock LLM provider. " +
              `Install it with: npm install @aws-sdk/client-bedrock-runtime (original error: ${detail})`,
          );
        },
      );
    }
    return this.sdkPromise;
  }

  /** Memoized Bedrock client; an injected `config.client` short-circuits the SDK. */
  private async getClient(): Promise<any> {
    if (this.clientOverride) return this.clientOverride;
    if (!this.clientPromise) {
      this.clientPromise = this.getSDK().then(
        ({ BedrockRuntimeClient }) =>
          new BedrockRuntimeClient(this.clientConfig),
      );
    }
    return this.clientPromise;
  }

  /**
   * Split messages into a top-level `system` block (Converse passes system
   * prompts separately) and role-tagged content blocks for everything else.
   */
  private formatMessages(messages: Message[]): {
    system?: { text: string }[];
    converseMessages: { role: string; content: { text: string }[] }[];
  } {
    const systemParts: string[] = [];
    const converseMessages: { role: string; content: { text: string }[] }[] =
      [];

    for (const msg of messages) {
      const role = msg.role;
      const content =
        typeof msg.content === "string"
          ? msg.content
          : JSON.stringify(msg.content);
      if (role === "system") {
        systemParts.push(content);
      } else {
        converseMessages.push({
          role: role === "assistant" ? "assistant" : "user",
          content: [{ text: content }],
        });
      }
    }

    if (converseMessages.length === 0) {
      converseMessages.push({ role: "user", content: [{ text: "" }] });
    }

    return {
      system: systemParts.length
        ? [{ text: systemParts.join("\n") }]
        : undefined,
      converseMessages,
    };
  }

  /**
   * Build the Converse `inferenceConfig`. Anthropic and MiniMax reasoning
   * models reject requests carrying both `temperature` and `topP`, so `topP`
   * is omitted for those families (mirrors the Python `_build_inference_config`).
   */
  private buildInferenceConfig(): Record<string, any> {
    const inferenceConfig: Record<string, any> = {
      maxTokens: this.maxTokens,
      temperature: this.temperature,
    };
    if (
      this.topP != null &&
      !["anthropic", "minimax"].includes(this.provider)
    ) {
      inferenceConfig.topP = this.topP;
    }
    return inferenceConfig;
  }

  /** Convert OpenAI-style tools to the Converse `toolConfig` shape. */
  private convertToolsToConverse(tools: any[]): any | undefined {
    if (!tools || tools.length === 0) return undefined;
    const converseTools = tools
      .filter((t) => t?.type === "function" && t.function)
      .map((t) => ({
        toolSpec: {
          name: t.function.name,
          description: t.function.description || "",
          inputSchema: { json: t.function.parameters || {} },
        },
      }));
    return converseTools.length ? { tools: converseTools } : undefined;
  }

  /**
   * Build a provider-family-specific request body for the legacy InvokeModel
   * path. Unlike Converse, each model family has its own body shape. Mirrors
   * the per-family branches in the Python provider (`mem0/llms/aws_bedrock.py`).
   */
  private buildInvokeBody(messages: Message[]): Record<string, any> {
    const systemParts: string[] = [];
    const turns: { role: string; content: string }[] = [];
    for (const msg of messages) {
      const content =
        typeof msg.content === "string"
          ? msg.content
          : JSON.stringify(msg.content);
      if (msg.role === "system") systemParts.push(content);
      else
        turns.push({
          role: msg.role === "assistant" ? "assistant" : "user",
          content,
        });
    }
    const system = systemParts.join("\n");
    const prompt = turns.map((t) => `${t.role}: ${t.content}`).join("\n");

    switch (this.provider) {
      case "anthropic":
        return {
          anthropic_version: "bedrock-2023-05-31",
          max_tokens: this.maxTokens,
          temperature: this.temperature,
          ...(this.topP != null ? { top_p: this.topP } : {}),
          ...(system ? { system } : {}),
          messages: turns.map((t) => ({
            role: t.role,
            content: [{ type: "text", text: t.content }],
          })),
        };
      case "amazon":
      case "titan":
        return {
          inputText: `${system ? system + "\n" : ""}${prompt}`,
          textGenerationConfig: {
            maxTokenCount: this.maxTokens,
            temperature: this.temperature,
            ...(this.topP != null ? { topP: this.topP } : {}),
          },
        };
      case "meta":
      case "llama":
        return {
          prompt: `${system ? system + "\n" : ""}${prompt}`,
          max_gen_len: this.maxTokens,
          temperature: this.temperature,
          ...(this.topP != null ? { top_p: this.topP } : {}),
        };
      case "mistral":
        return {
          prompt: `${system ? system + "\n" : ""}${prompt}`,
          max_tokens: this.maxTokens,
          temperature: this.temperature,
          ...(this.topP != null ? { top_p: this.topP } : {}),
        };
      case "cohere":
        return {
          message: prompt,
          max_tokens: this.maxTokens,
          temperature: this.temperature,
          ...(this.topP != null ? { p: this.topP } : {}),
        };
      case "ai21":
      case "j2":
        return {
          prompt: `${system ? system + "\n" : ""}${prompt}`,
          maxTokens: this.maxTokens,
          temperature: this.temperature,
          ...(this.topP != null ? { topP: this.topP } : {}),
        };
      default:
        throw new Error(
          `InvokeModel body not implemented for provider '${this.provider}'. ` +
            `Use the default Converse API for this model, or add a body builder for '${this.provider}'.`,
        );
    }
  }

  /** Parse a provider-family-specific InvokeModel response body into text. */
  private parseInvokeResponse(body: any): string {
    switch (this.provider) {
      case "anthropic":
        return body?.content?.[0]?.text ?? body?.completion ?? "";
      case "amazon":
      case "titan":
        return body?.results?.[0]?.outputText ?? "";
      case "meta":
      case "llama":
        return body?.generation ?? "";
      case "mistral":
        return body?.outputs?.[0]?.text ?? "";
      case "cohere":
        return body?.text ?? body?.generations?.[0]?.text ?? "";
      case "ai21":
      case "j2":
        return body?.completions?.[0]?.data?.text ?? "";
      default:
        return "";
    }
  }

  /** Legacy InvokeModel path for models not covered by Converse (#6023). */
  private async invokeModel(messages: Message[]): Promise<string> {
    const body = this.buildInvokeBody(messages);
    const [{ InvokeModelCommand }, client] = await Promise.all([
      this.getSDK(),
      this.getClient(),
    ]);
    const response = await client.send(
      new InvokeModelCommand({
        modelId: this.model,
        contentType: "application/json",
        accept: "application/json",
        body: JSON.stringify(body),
      }),
    );
    // Response body is a Uint8Array in the real SDK.
    const raw = response?.body;
    let parsed: any = raw;
    if (raw instanceof Uint8Array || Buffer.isBuffer(raw)) {
      parsed = JSON.parse(Buffer.from(raw).toString("utf8"));
    } else if (typeof raw === "string") {
      parsed = JSON.parse(raw);
    } else if (raw && typeof raw === "object" && "transformToString" in raw) {
      // Some SDK versions expose a streaming blob with transformToString.
      parsed = JSON.parse(await (raw as any).transformToString());
    }
    return this.parseInvokeResponse(parsed);
  }

  private async converse(messages: Message[], tools?: any[]): Promise<any> {
    const { system, converseMessages } = this.formatMessages(messages);
    const input: Record<string, any> = {
      modelId: this.model,
      messages: converseMessages,
      inferenceConfig: this.buildInferenceConfig(),
    };
    if (system) input.system = system;
    const toolConfig = tools ? this.convertToolsToConverse(tools) : undefined;
    if (toolConfig) input.toolConfig = toolConfig;

    const [{ ConverseCommand }, client] = await Promise.all([
      this.getSDK(),
      this.getClient(),
    ]);
    return client.send(new ConverseCommand(input));
  }

  /** Pull the first text block out of a Converse response. */
  private parseText(response: any): string {
    const content = response?.output?.message?.content || [];
    for (const block of content) {
      if (block && typeof block.text === "string") return block.text;
    }
    return "";
  }

  /** Collect any toolUse blocks out of a Converse response. */
  private parseToolCalls(response: any): { name: string; arguments: string }[] {
    const content = response?.output?.message?.content || [];
    const calls: { name: string; arguments: string }[] = [];
    for (const block of content) {
      if (block?.toolUse) {
        calls.push({
          name: block.toolUse.name,
          arguments: JSON.stringify(block.toolUse.input ?? {}),
        });
      }
    }
    return calls;
  }

  async generateResponse(
    messages: Message[],
    _responseFormat?: { type: string },
    tools?: any[],
  ): Promise<string | LLMResponse> {
    try {
      // Legacy InvokeModel path does not support Converse-style tool config.
      if (this.api === "invoke_model") {
        return await this.invokeModel(messages);
      }
      const response = await this.converse(messages, tools);
      if (tools && tools.length) {
        const toolCalls = this.parseToolCalls(response);
        if (toolCalls.length) {
          return {
            content: this.parseText(response),
            role: "assistant",
            toolCalls,
          };
        }
      }
      return this.parseText(response);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      throw new Error(`AWS Bedrock LLM failed: ${message}`);
    }
  }

  async generateChat(messages: Message[]): Promise<LLMResponse> {
    try {
      const response = await this.converse(messages);
      return { content: this.parseText(response), role: "assistant" };
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      throw new Error(`AWS Bedrock LLM failed: ${message}`);
    }
  }
}
